src/devices/Lcm1602a1/Lcm1602a1.cs

```diff
@@ -0,0 +1,577 @@
+        /// Writes a value to a pin.
+        /// </summary>
+        /// <param name="pinNumber">The pin number.</param>
+        /// <param name="value">The value that needs to be written.</param>
+        private void Write(int pinNumber, PinValue value)
+        {
+            if (_usingMcp)
+            {
+                _mcpController.WritePin(pinNumber, value);
+            }
+            else
+            {
+                _controller.Write(pinNumber, value);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_controller != null)
+            {
+                _controller.Dispose();
+                _controller = null;
+            }
+        }
+
+        /// <summary>
+        /// Initializes the display by setting the specified columns and lines.
+        /// </summary>
+        /// <param name="cols">The columns that will be used by the display.</param>
+        /// <param name="lines">The lines that will be used by the display.</param>
+        /// <param name="dotSize">The resolution that will be used. 5x8 is the default.</param>
+        public void Begin(byte cols, byte lines, DisplayFlags dotSize = DisplayFlags.LCD_5x8DOTS)
+        {
+            if (lines > 1)
+            {
+                _displayFunction |= DisplayFlags.LCD_2LINE;
+            }
+
+            _numLines = lines;
+
+            SetRowOffsets(0x00, 0x40, 0x14, 0x54);
+
+            // for some 1 line displays you can select a 10 pixel high font
+            if ((dotSize != DisplayFlags.LCD_5x8DOTS) && (lines == 1))
+            {
+                _displayFunction |= DisplayFlags.LCD_5x10DOTS;
+            }
+
+            _displayControl = DisplayFlags.LCD_DISPLAYON | DisplayFlags.LCD_CURSOROFF | DisplayFlags.LCD_BLINKOFF;
+            _displayMode = DisplayFlags.LCD_ENTRYLEFT | DisplayFlags.LCD_ENTRYSHIFTDECREMENT;
+
+            SetPinMode(_rsPin, PinMode.Output);
+            // we can save 1 pin by not using RW. Indicate by passing null instead of a pin
+            if (_rwPin != -1)
+            {
+                SetPinMode(_rwPin, PinMode.Output);
+            }
+            if (_backlight != -1)
+            {
+                SetPinMode(_backlight, PinMode.Output);
+                Write(_backlight, PinValue.High);
+            }
+            SetPinMode(_enablePin, PinMode.Output);
+
+            // Do this just once, instead of every time a character is drawn (for speed reasons).
+            for (int i = 0; i < _dataPins.Length; ++i)
+            {
+                SetPinMode(_dataPins[i], PinMode.Output);
+            }
+
+            // SEE PAGE 45/46 FOR INITIALIZATION SPECIFICATION!
+            // according to datasheet, we need at least 40ms after power rises above 2.7V
+            // before sending commands. Arduino can turn on way before 4.5V so we'll wait 50
+            DelayMicroseconds(50000);
+            
+            // Initialize the display.
+            Write((byte)0x33);
```

Comment:
I'm having a hard time tying this back to the datasheet. It looks like we were trying to reset the display if I'm following page 45 correctly. Is this necessary? If so, it looks like you need to send 00110000 three times with waits after the first two (4.1ms and 100μs) (without checking BF).  This appears to me like we're not really resetting?

On page 27 it says function set instruction must be run before all other instructions. On page 40 it shows function set before display control and entry mode.

I went back and checked the samples again on the CrowPi and I think I'm seeing behavior that backs this up. If I start up from power off the code doesn't work and prints garbage. If I then run the CrowPi python sample the display gets reset correctly and the sample works from that point. When I tested before I didn't power off the CrowPi.